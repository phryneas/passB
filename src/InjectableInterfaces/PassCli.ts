export interface PassCli {
  list(): Promise<string[]>;
  show(path: string): Promise<string[]>;
  getHostAppVersion(): Promise<string>;
}
