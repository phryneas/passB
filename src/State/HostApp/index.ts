import {createSelector} from 'reselect';
import {actionCreatorFactory} from 'typescript-fsa';
import {reducerWithInitialState} from 'typescript-fsa-reducers';
import {StoreContents} from 'InjectableInterfaces/State';
import {getHostApState} from 'State/Selectors';

// interfaces
export interface Error {
  message: string;
  type: ErrorType;
}

export interface HostAppState {
  version: string;
  lastError?: Error;
}

export enum ErrorType {
  PASS_EXECUTION_ERROR = 'PASS_EXECUTION_ERROR',
  HOST_APP_ERROR = 'HOST_APP_ERROR',
}

// actions
const actionCreator = actionCreatorFactory('HOST_APP');
export const setVersion = actionCreator<string>('SET_VERSION');
export const setLastError = actionCreator<Error>('SET_LAST_ERROR');
export const clearLastError = actionCreator<void>('CLEAR_LAST_ERROR');

export const Actions = {
  setVersion,
  setLastError,
  clearLastError,
};

// reducer
export const initialState = {
  version: 'unknown',
};
export const reducer = reducerWithInitialState(initialState)
  .case(setVersion, (state: HostAppState, version: string): HostAppState => ({...state, version}))
  .case(setLastError, (state: HostAppState, lastError: Error): HostAppState => ({...state, lastError}))
  .case(clearLastError, ({lastError, ...stateRest}: HostAppState): HostAppState => stateRest)
  .build();

// selectors
export const getLastError: (state: StoreContents) => Error | undefined =
  createSelector(getHostApState, (state: HostAppState) => state.lastError);

export const getVersion: (state: StoreContents) => string =
  createSelector(getHostApState, (state: HostAppState) => state.version);
