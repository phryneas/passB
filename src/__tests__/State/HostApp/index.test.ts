import {getLastError, getVersion, reducer, Actions, ErrorType} from 'State/HostApp';
import {StoreContents} from '../../../InjectableInterfaces/State';

describe('HostApp State', () => {
  test('setLastError', () => {
    expect(reducer(undefined!, Actions.setLastError({type: ErrorType.HOST_APP_ERROR, message: 'test'}))).toMatchSnapshot();
  });

  test('clearLastError', () => {
    const before = reducer(undefined!, Actions.setLastError({type: ErrorType.HOST_APP_ERROR, message: 'test'}));
    expect(reducer(before, Actions.clearLastError(void 0))).toMatchSnapshot();
  });

  test('getLastError', () => {
    const state = {hostApp: reducer(undefined!, Actions.setLastError({type: ErrorType.HOST_APP_ERROR, message: 'test'}))};
    expect(getLastError(state as StoreContents)).toEqual({type: ErrorType.HOST_APP_ERROR, message: 'test'});
  });

  test('setVersion', () => {
    expect(reducer(undefined!, Actions.setVersion('1.2.3'))).toMatchSnapshot();
  });

  test('getVersion', () => {
    const state = {hostApp: reducer(undefined!, Actions.setVersion('1.2.3'))};
    expect(getVersion(state as StoreContents)).toEqual('1.2.3');
  });
});
