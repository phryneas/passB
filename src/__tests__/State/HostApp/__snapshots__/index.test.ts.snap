// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HostApp State clearLastError 1`] = `
Object {
  "version": "unknown",
}
`;

exports[`HostApp State setLastError 1`] = `
Object {
  "lastError": Object {
    "message": "test",
    "type": "HOST_APP_ERROR",
  },
  "version": "unknown",
}
`;

exports[`HostApp State setVersion 1`] = `
Object {
  "version": "1.2.3",
}
`;
